src/map/ai/states/range_state.cpp

```diff
@@ -53,7 +54,10 @@ CRangeState::CRangeState(CBattleEntity* PEntity, uint16 targid)
     }
 
     auto delay = m_PEntity->GetRangedWeaponDelay(false);
-    delay      = battleutils::GetSnapshotReduction(m_PEntity, delay);
+    // save the initial dmg and delay of weapon for checking later in the state
+    m_initialDamage = PEntity->GetRangedWeaponDmg();
+    m_initialDelay  = delay;
```

Comment:
Are players swapping between the same item? Surely storing the item ID is the correct answer for this otherwise?

Similarly, this seems remedial, rather than preventative. Would it make more sense to check where we do gear changes and interrupt the user's state if they change the relevant weapon - rather than in the state itself?

Reply:
Also, a sufficiently quick player with a sufficiently slow weapon will be able to swap from set A, to set B, back to set A and have the ranged attack continue - I don't have any proof, but I can only assume that isn't what retail does, so this should probably be handled where we handle gear changing.

Reply:
Hey guys. I tried this and retail will actually prevent the swap while mid-ranged attack and sometimes do the swap ***after*** if you did it late enough. A similar thing happens if you try to change your ring equip in the middle of using a warp ring - it'll ignore you unless it was near completion, then it'll just do it _after_ instead of during. I presume because the packet then arrived after server side "unlocked" it. You just plain can't swap till server deems it safe.

Reply:
> Hey guys. I tried this and retail will actually prevent the swap while mid-ranged attack and sometimes do the swap _**after**_ if you did it late enough. A similar thing happens if you try to change your ring equip in the middle of using a warp ring - it'll ignore you unless it was near completion, then it'll just do it _after_ instead of during. I presume because the packet then arrived after server side "unlocked" it. You just plain can't swap till server deems it safe.

Thanks for retail info, I have updated PR to prevent such gear changing if the player is in a range state, thus should be fairly retail accurate.